stan/math/gpu/cholesky_decompose.hpp

```diff
@@ -61,56 +44,87 @@ inline matrix_gpu cholesky_decompose_recursion(matrix_gpu& A,
     // or calls the kernel  directly if the size of the block is small enough
-    matrix_gpu L_11
-        = stan::math::internal::cholesky_decompose_recursion(A_11, min_block);
+    matrix_gpu L_11 = stan::math::internal::cholesky_decompose(A_11, min_block);
     // Copies L_11 back to the input matrix
-    A.sub_block(L_11, 0, 0, offset, offset, block, block);
-    // Copies a block of the input A into A_21
-    auto block_offset = offset + block;
-    A_21.sub_block(A, block_offset, offset, 0, 0, block_subset, block);
+    A.sub_block(L_11, 0, 0, 0, 0, block, block);
+
+    auto block_subset = A.rows() - block;
+    matrix_gpu A_21(block_subset, block);
+    A_21.sub_block(A, block, 0, 0, 0, block_subset, block);
     // computes A_21*((L_11^-1)^T)
-    // and copies the resulting submatrix to the input matrix
+    // and copies the resulting submatrix to the right hand corner of A
     matrix_gpu L_21 = A_21 * transpose(lower_triangular_inverse(L_11));
-    A.sub_block(L_21, 0, 0, block_offset, offset, block_subset, block);
-    A_22.sub_block(A, block_offset, block_offset, 0, 0, block_subset,
-                   block_subset);
+    A.sub_block(L_21, 0, 0, block, 0, block_subset, block);
+    matrix_gpu A_22(block_subset, block_subset);
+    A_22.sub_block(A, block, block, 0, 0, block_subset, block_subset);
     // computes A_22 - L_21*(L_21^T)
     matrix_gpu L_22 = A_22 - multiply_transpose(L_21);
-    A.sub_block(L_22, 0, 0, block_offset, block_offset, block_subset,
-                block_subset);
-    offset += block;
+    // copy L_22 into A's left hand corner
+    A.sub_block(L_22, 0, 0, block, block, block_subset, block_subset);
+    // Computes the Cholesky factor for the remaining submatrix
+    matrix_gpu A_rem_11(block_subset, block_subset);
+    A_rem_11.sub_block(A, block, block, 0, 0, block_subset, block_subset);
+    matrix_gpu L_rem_11
+        = stan::math::internal::cholesky_decompose(A_rem_11, min_block);
+    A.sub_block(L_rem_11, 0, 0, block, block, block_subset, block_subset);
+    check_nan("cholesky_decompose_gpu", "Matrix m", A);
+    check_diagonal_zeros("cholesky_decompose_gpu", "Matrix m", A);
+    A.zeros<stan::math::TriangularViewGPU::Upper>();
+    return A;
   }
-  // Computes the Cholesky factor for the remaining submatrix
-  const auto remaining_rows = A.rows() - offset;
-  if (remaining_rows > 0) {
-    matrix_gpu A_11(remaining_rows, remaining_rows);
-    A_11.sub_block(A, offset, offset, 0, 0, remaining_rows, remaining_rows);
-    // calculate the cholesky factor for the remaining part of the matrix
-    matrix_gpu L_11
-        = stan::math::internal::cholesky_decompose_recursion(A_11, min_block);
-    A.sub_block(L_11, 0, 0, offset, offset, remaining_rows, remaining_rows);
-  }
-  check_nan("cholesky_decompose_gpu", "Matrix m", A);
-  check_diagonal_zeros("cholesky_decompose_gpu", "Matrix m", A);
-  A.zeros<stan::math::TriangularViewGPU::Upper>();
+  // This should never happen
   return A;
```

Comment:
Yep! I saw you do that once somewhere else and I'm impressionable. Removing the else here works just fine and looks cleaner

Reply:
Fair enough. I think in cases like this where the else block is almost the entirety of the function and the first couple ifs return early, it's better to just omit the else so that someone reading it doesn't have to figure out why it should never happen to double check it or whatever. I should have done the Python illustration that way too but it was so short that it didn't really matter there. Oops.